### DISCUSSION TITLE
Package Name Redundancy Doubt

### DISCUSSION THREAD

**Discussion by chessmadridista** [fastapi/fastapi#10269] [Questions] [answered]
In ```fastapi/__init__.py```, many packages are imported as follows:
```
import module_name as module_name
```
How is it better than simply writing 
```
import module_name
```
?

And how do I create an issue out of it (for refactoring purposes) if there are no major benefits?

**Comment by n8sty** [comment]
This is a quirk of FastAPI and a choice on the part of the package author. If you believe this is problematic for some reason try opening a PR that fixes the problem you see and it's probably worth including a somewhat lengthy explanation about why this is actually a problem. In reality, I don't think this hypothetical change gets merged.

**Comment by nymous** [comment, marked as answer]
This is probably because it is the way to mark an import as explicitly reexported for type checkers.
If you don't do this (or use the `__all__` syntax), type checkers would complain with an error similar to this:
> Module "fastapi" does not explicitly export attribute "module_name"; implicit reexport disabled


See [Mypy documentation](https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-no-implicit-reexport)
> By default, imported values to a module are treated as exported and mypy allows other modules to import them. This flag changes the behavior to not re-export unless the item is imported using from-as or is included in `__all__`. Note this is always treated as enabled for stub files. For example:
> ```py
> # This won't re-export the value
> from foo import bar
> 
> # Neither will this
> from foo import bar as bang
> 
> # This will re-export it as bar and allow other modules to import it
> from foo import bar as bar
> 
> # This will also re-export bar
> from foo import bar
> __all__ = ['bar']
> ```

and [Pyright documentation](https://github.com/microsoft/pyright/blob/main/docs/typed-libraries.md#library-interface)
> Each module exposes a set of symbols. Some of these symbols are considered “private” — implementation details that are not part of the library’s interface. Type checkers like pyright use the following rules to determine which symbols are visible outside of the package.
> 
> - Symbols whose names begin with an underscore (but are not dunder names) are considered private.
> - Imported symbols are considered private by default. If they use the “import A as A” (a redundant module alias), “from X import A as A” (a redundant symbol alias), or “from . import A” forms, symbol “A” is not private unless the name begins with an underscore. If a file `__init__.py` uses the form “from .A import X”, symbol “A” is not private unless the name begins with an underscore (but “X” is still private). If a wildcard import (of the form “from X import *”) is used, all symbols referenced by the wildcard are not private.
> - A module can expose an `__all__` symbol at the module level that provides a list of names that are considered part of the interface. The `__all__` symbol indicates which symbols are included in a wildcard import. All symbols included in the `__all__` list are considered public even if the other rules above would otherwise indicate that they were private. For example, this allows symbols whose names begin with an underscore to be included in the interface.
